src/hash.c

```diff
@@ -261,25 +289,78 @@ void hash_fill(uint32_t* p, R_len_t size, SEXP x) {
     p[i] = hash_combine(p[i], HASHER(xp));      \
   }
 
-static inline void lgl_hash_fill(uint32_t* p, R_len_t size, SEXP x) {
+#define HASH_FILL_NA_PROPAGATE(CTYPE, CONST_DEREF, HASHER, NA_VALUE)    \
+  const CTYPE* xp = CONST_DEREF(x);                                     \
+                                                                        \
+  for (R_len_t i = 0; i < size; ++i, ++xp) {                            \
+    if (p[i] == HASH_MISSING) {                                         \
+      continue;                                                         \
+    }                                                                   \
+    if (*xp == NA_VALUE) {                                              \
+      p[i] = HASH_MISSING;                                              \
+    } else {                                                            \
+      p[i] = hash_combine(p[i], HASHER(xp));                            \
```

Comment:
There are two `p[i]` accesses here that could be stored in a single `const uint32_t hash = p[i]`. This might give a small performance boost.

Same for `HASH_FILL_NA_PROPAGATE_CMP` and `HASH_FILL_BARRIER_NA_PROPAGATE`